proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,747 @@
+presentations of dependent types with linearity-in-kinds disallow
+linear types as arguments of dependent functions.
+
+Future work
+~~~~~~~~~~~
+
+Something that hasn't been touched up by this proposal is the idea of
+declaring toplevel linear binders
+
+::
+
+  module Foo where
+  token ::('1) A  -- made up syntax
+
+Here ``token`` would have be consumed exactly once by the program,
+this property is a link-time property. This generalised the
+``RealWorld`` token which is currently magically inserted in the
+``main`` function (the existence of which is checked at link time).
+
+This would allow libraries to abstract on ``main`` or to provide their
+own linearly-threaded token.
+
+Unresolved questions
+--------------------
+
+Inference
+~~~~~~~~~
+
+- There is no systematic account of type inference. Can it be made
+  predictable when a type annotation is required? For compatibility
+  reasons, we want to infer unrestricted arrows conservatively, but
+  experience shows that it can result in very surprising type errors.
+
+- In the formalism, case expressions are indexed by a multiplicity:
+  ``case_p`` (and similarly ``let_p``). In the surface language, we
+  can deduce the multiplicity in equations when their is a type
+  annotation.
+
+  ::
+
+    fst :: (a,b) -> a
+    fst (a,_) = a    -- this is inferred as a case_ω
+
+    swap :: (a,b) ->. (b,a)
+    swap (a,b) = (b,a)   -- this is inferred as a case_1
+
+  But what of explicit ``case`` and ``let`` in the surface language? We
+  can annotate them with a multiplicity, but it is generally clear from
+  the context which multiplicity is meant. So the multiplicity
+  annotation really ought to be inferred. The general idea is: if
+  their is any linear variable in the scrutiny, then the case must be
+  linear, and if there are only unrestricted variables, it can be
+  unrestricted. Is it sound to always pick the highest possible value ?
+  What if there are multiplicities with variable multiplicity ?
+
+Syntax
+~~~~~~
+
+TODO: Overloading do for ``IOL`` (local vs too global ``-XRebindableSyntax``).
+
+Formalism
+~~~~~~~~~
+
+TODO: probably ought to be moved to the implementation plan section:
+
+There's one thing I papered over on the formalism: in Core, ``case``
+is of the form ``case u as x of { <alternatives> }`` where ``x``
+represents the head normal form of ``u``. It is a widely use tool in
+Core to Core passes. It is in particular used to implement the default
+alternative is a case:
+
+::
+
+  fmap' :: (a -> a) -> Maybe a -> Maybe a
+  fmap' (Just x) = Just (f x)
+  fmap' y = y
+
+is elaborated into
+
+::
+
+  \f o -> case o as y of { Just x -> Just (f x) ; WILDCARD -> y }
+
+But it is not obvious what to do for linear cases. The following is a
+linearity violation as ``y`` in a sense contains ``x`` (basically, you
+could define a function ``a ->. (a,a)`` generically with this).
+
+::
+  case_1 o as y of { Just x -> Just (x,y) }
```

Comment:
What about treating `y` as `Cj x1 ... xn` in each alternative of the form `C x1 ... xn  -> ... y ...`?

Reply:
@aspiwack, in which case does it make sense to use `y` in a non-default alternative?

Reply:
in `x@(Just _)` for instance.

Reply:
Is it ok then to pattern match multiple times on a linear value?
```Haskell
case x of
  Just _ -> case x of
    Just (Just _)  -> case x of
      Just (Just (Just _)) -> ...
```

Reply:
Well, it is not easy to make this type-check, but conceptually, it is a correct program.

Though now that I think about it, it makes self-deallocating heap object more complex to handle. It may be that the best answer is to treat `case` conservatively, and not these programs.